File: tt-train/sources/ttml/modules/distributed/linear.cpp
// SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
//
// SPDX-License-Identifier: Apache-2.0

#include "linear.hpp"

#include <cmath>

#include "autograd/auto_context.hpp"
#include "autograd/tensor.hpp"
#include "core/tt_tensor_utils.hpp"
#include "init/cpu_initializers.hpp"
#include "init/tensor_initializers.hpp"
#include "ops/binary_ops.hpp"
#include "ops/distributed/comm_ops.hpp"
#include "ops/linear_op.hpp"

namespace ttml::modules::distributed {

RowParallelLinear::RowParallelLinear(
    uint32_t in_features,
    uint32_t out_features,
    bool has_bias,
    bool input_is_parallel,
    std::optional<uint32_t> shard_dim) :
    m_input_is_parallel(input_is_parallel), /* input is parallel across TP axis */
    m_shard_dim(shard_dim) /* shard dimension in the device mesh */ {
    initialize_tensors(in_features, out_features, has_bias);

    create_name("row_parallel_linear");
    register_tensor(m_weight, "weight");
    if (m_bias != nullptr) {
        register_tensor(m_bias, "bias");
    }
}

autograd::TensorPtr RowParallelLinear::operator()(const autograd::TensorPtr& tensor) {
    auto x = tensor;
    if (!m_input_is_parallel) {
        x = ops::distributed::scatter(x, tensor->get_rank() - 1U, m_shard_dim);
    }
    // do not pass bias
    x = ops::linear_op(x, m_weight, /* bias */ nullptr);

    /*
        All reduce with noop backward to avoid double all reduce in backward pass. This happens due to broadcast (no op in forward pass)
        does all reduce in backward pass. See similar implementation in fairscale for more details.
        https://github.com/facebookresearch/fairscale/blob/main/fairscale/nn/model_parallel/mappings.py#L102
    */
    x = ops::distributed::all_reduce(x, /* noop_backward */ m_input_is_parallel, m_shard_dim);
    if (m_bias != nullptr) {
        x = ops::add(x, m_bias);
    }
    return x;
}

void RowParallelLinear::initialize_tensors(uint32_t in_features, uint32_t out_features, bool has_bias) {
    auto* device = &autograd::ctx().get_device();
    auto mesh_shape = device->shape();

    // Determine TP size based on mesh shape and shard_dim
    uint32_t tp_size = static_cast<uint32_t>(device->num_devices());
    if (m_shard_dim.has_value() && mesh_shape.dims() == 2) {
        // For 2D mesh with explicit shard_dim: TP size is mesh dimension specified by shard_dim
        tp_size = mesh_shape[m_shard_dim.value()];
    }

    if (in_features % tp_size != 0) {
        throw std::runtime_error(fmt::format(
            "Input features must be divisible by the TP size. Input features = {}, TP size = {}",
            in_features,
            tp_size));
    }

    const auto weight_shape = ttnn::Shape({1, 1, out_features, in_features});
    uint32_t rank = 4U;
    const float init_k = std::sqrt(1.F / static_cast<float>(in_features));

    auto weight = init::uniform_init(weight_shape, init::UniformRange{-init_k, init_k});

    const auto mapper = ttnn::distributed::shard_tensor_to_mesh_mapper(*device, rank - 1U, m_shard_dim);
    m_weight = autograd::create_tensor(
        ttml::core::from_xtensor<float, ttnn::DataType::BFLOAT16>(weight, device, ttnn::Layout::TILE, mapper.get()));

    if (has_bias) {
        auto bias_shape = ttnn::Shape({1, 1, 1, out_features});
        m_bias = ttml::autograd::create_tensor();
        init::uniform_init(m_bias, bias_shape, init::UniformRange{-init_k, init_k});
    }
}

ColumnParallelLinear::ColumnParallelLinear(
    uint32_t in_features,
    uint32_t out_features,
    bool has_bias,
    bool gather_output,
    std::optional<uint32_t> shard_dim) :
    m_gather_output(gather_output),
    m_shard_dim(shard_dim) {
    initialize_tensors(in_features, out_features, has_bias);

    create_name("column_parallel_linear");
    register_tensor(m_weight, "weight");
    if (m_bias != nullptr) {
        register_tensor(m_bias, "bias");
    }
}

autograd::TensorPtr ColumnParallelLinear::operator()(const autograd::TensorPtr& tensor) {
    auto x = tensor;
    // Broadcast data along TP dimension to ensure all TP devices in each DP group have the same data
    x = ops::distributed::broadcast(x, m_shard_dim);
    x = ops::linear_op(x, m_weight, m_bias);
    if (m_gather_output) {
        // All-gather output along TP dimension to gather sharded outputs within each DP group
        x = ops::distributed::all_gather(x, tensor->get_rank() - 1U, m_shard_dim);
    }
    return x;
}

void ColumnParallelLinear::initialize_tensors(uint32_t in_features, uint32_t out_features, bool has_bias) {
    auto* device = &autograd::ctx().get_device();
    auto mesh_shape = device->shape();

    // Determine TP size based on mesh shape and shard_dim
    uint32_t tp_size = 1U;
    if (m_shard_dim.has_value() && mesh_shape.dims() == 2) {
        // For 2D mesh with explicit shard_dim: TP size is mesh dimension specified by shard_dim
        tp_size = mesh_shape[m_shard_dim.value()];
    } else {
        // 1D mesh: use all devices
        tp_size = static_cast<uint32_t>(device->num_devices());
    }

    if (out_features % tp_size != 0) {
        throw std::runtime_error(fmt::format(
            "Output features must be divisible by the TP size. Output features = {}, TP size = {}",
            out_features,
            tp_size));
    }

    auto weight_shape = ttnn::Shape({1, 1, out_features, in_features});
    uint32_t rank = 4U;
    const float init_k = std::sqrt(1.F / static_cast<float>(in_features));

    auto weight = init::uniform_init(weight_shape, init::UniformRange{-init_k, init_k});

    const auto weight_mapper = ttnn::distributed::shard_tensor_to_mesh_mapper(*device, rank - 2U, m_shard_dim);
    m_weight = autograd::create_tensor(
        ttml::core::from_xtensor<float, ttnn::DataType::BFLOAT16>(weight, device, ttnn::Layout::TILE, weight_mapper.get()));

    if (has_bias) {
        const auto bias_shape = ttnn::Shape({1, 1, 1, out_features});
        const auto bias = init::uniform_init(bias_shape, init::UniformRange{-init_k, init_k});
        const auto bias_mapper = ttnn::distributed::shard_tensor_to_mesh_mapper(*device, rank - 1U, m_shard_dim);
        m_bias = autograd::create_tensor(
            ttml::core::from_xtensor<float, ttnn::DataType::BFLOAT16>(bias, device, ttnn::Layout::TILE, bias_mapper.get()));
    }
}

}  // namespace ttml::modules::distributed

File: tt-train/sources/ttml/ttml/modules/linear.py
# SPDX-FileCopyrightText: © 2026 Tenstorrent USA, Inc.
#
# SPDX-License-Identifier: Apache-2.0

from typing import Callable
import math

import ml_dtypes

import ttnn
import ttml

from ttml.parallel import SEQUENCE_DIM, mark_sequence_parallel

from .module_base import AbstractModuleBase
from .parameter import Parameter


def column_parallel_input(x, cluster_axis: int, sequence_parallel: bool):
    """The input of a column-parallel matmul: the full sequence, replicated on every TP rank."""
    if sequence_parallel:
        return ttml.ops.distributed.all_gather(
            x, SEQUENCE_DIM, cluster_axis, ttml.ops.distributed.GradOutputType.SHARDED
        )
    return ttml.ops.distributed.broadcast(x, cluster_axis)


def row_parallel_output(x, cluster_axis: int, sequence_parallel: bool, input_is_parallel: bool):
    """The output of a row-parallel matmul: the partial products summed across TP ranks."""
    if sequence_parallel:
        return ttml.ops.distributed.reduce_scatter(x, SEQUENCE_DIM, cluster_axis)
    return ttml.ops.distributed.all_reduce(x, input_is_parallel, cluster_axis)


class LinearLayer(AbstractModuleBase):
    """Fully-connected linear layer: y = x @ W^T + b."""

    def __init__(
        self,
        in_features: int,
        out_features: int,
        has_bias: bool = True,
        weight_init: Callable | None = None,
        bias_init: Callable | None = None,
    ) -> None:
        super().__init__()

        self.in_features = in_features
        self.out_features = out_features

        if weight_init is None:
            k = math.sqrt(1.0 / in_features)
            weight_init = ttml.init.uniform(-k, k)
        if bias_init is None:
            k = math.sqrt(1.0 / in_features)
            bias_init = ttml.init.uniform(-k, k)

        weight_shape = (1, 1, out_features, in_features)
        self.weight = Parameter(weight_init(weight_shape))

        if has_bias:
            bias_shape = (1, 1, 1, out_features)
            self.bias = Parameter(bias_init(bias_shape))
        else:
            self.bias = None

    def __reduce__(self):
        return (
            self.__class__,
            (self.in_features, self.out_features, self.bias is not None),
            self.__getstate__(),
        )

    def __getstate__(self):
        return {
            "weight": self.weight.tensor.to_numpy(ttnn.DataType.FLOAT32),
            "bias": self.bias.tensor.to_numpy(ttnn.DataType.FLOAT32) if self.bias is not None else None,
        }

    def __setstate__(self, state):
        self.weight.tensor.set_value(
            ttml.autograd.Tensor.from_numpy(
                state["weight"].astype(ml_dtypes.bfloat16), layout=ttnn.Layout.TILE
            ).get_value()
        )
        if state["bias"] is not None:
            if self.bias is None:
                raise ValueError("LinearLayer bias was improperly initialized when deserializing from Pickle")
            self.bias.tensor.set_value(
                ttml.autograd.Tensor.from_numpy(
                    state["bias"].astype(ml_dtypes.bfloat16), layout=ttnn.Layout.TILE
                ).get_value()
            )

    def forward(self, x):
        """Compute linear projection of x."""
        bias_t = self.bias.tensor if self.bias is not None else None
        return ttml.ops.linear.linear(x, self.weight.tensor, bias_t)


class ColumnParallelLinear(AbstractModuleBase):
    """Column-parallel linear layer.

    The weight matrix ``(out_features, in_features)`` is sharded along the
    *output-feature* (row) dimension — i.e. dim 2 in the 4-D layout
    ``(1, 1, out_features, in_features)`` — so each TP device holds
    ``out_features // tp_size`` rows.  The bias is similarly sharded on dim 3.

    Args:
        gather_output: If ``True`` an all-gather is inserted after the matmul so
            the output is replicated across TP devices (needed when the consumer
            expects an un-sharded tensor, e.g. the final LM-head projection).
        sequence_parallel: If ``True`` the input arrives sharded along the sequence
            dimension across the TP axis (Megatron sequence parallelism). An
            all-gather over the sequence dim reconstructs the full sequence before
            the matmul.
        axis_name: Mesh axis used for tensor parallelism.
    """

    def __init__(
        self,
        in_features: int,
        out_features: int,
        has_bias: bool = True,
        weight_init: Callable | None = None,
        bias_init: Callable | None = None,
        gather_output: bool = False,
        sequence_parallel: bool = False,
        axis_name: str = "tp",
    ) -> None:
        super().__init__()

        self.in_features = in_features
        self.out_features = out_features
        self.gather_output = gather_output
        self.sequence_parallel = sequence_parallel
        self.axis_name = axis_name
        self.cluster_axis = ttml.mesh().axis_index(axis_name)

        if weight_init is None:
            k = math.sqrt(1.0 / in_features)
            weight_init = ttml.init.uniform(-k, k)
        if bias_init is None:
            k = math.sqrt(1.0 / in_features)
            bias_init = ttml.init.uniform(-k, k)

        weight_shape = (1, 1, out_features, in_features)
        mesh = ttml.mesh()
        weight_mapper = mesh.axis_mapper(axis_name, tdim=2)
        self.weight = Parameter(weight_init(weight_shape, mapper=weight_mapper))

        if has_bias:
            bias_shape = (1, 1, 1, out_features)
            bias_mapper = mesh.axis_mapper(axis_name, tdim=3)
            self.bias = Parameter(bias_init(bias_shape, mapper=bias_mapper))
        else:
            self.bias = None

    def forward(self, x):
        x = column_parallel_input(x, self.cluster_axis, self.sequence_parallel)
        bias_t = self.bias.tensor if self.bias is not None else None
        x = ttml.ops.linear.linear(x, self.weight.tensor, bias_t)
        if self.gather_output:
            # Reconstitute the full output across TP devices (dim 3 = last feature dim).
            x = ttml.ops.distributed.all_gather(x, 3, self.cluster_axis, ttml.ops.distributed.GradOutputType.REPLICATED)
        return x


class RowParallelLinear(AbstractModuleBase):
    """Row-parallel linear layer.

    The weight matrix ``(out_features, in_features)`` is sharded along the
    *input-feature* (column) dimension — i.e. dim 3 in the 4-D layout
    ``(1, 1, out_features, in_features)`` — so each TP device holds
    ``in_features // tp_size`` columns.  The bias is replicated (not sharded)
    because it is added after the all-reduce that sums partial products.

    Args:
        input_is_parallel: Set to ``True`` when the incoming activation is already
            sharded along the feature dimension (e.g. the output of a
            ``ColumnParallelLinear`` with ``gather_output=False``).  When ``False``
            a scatter is inserted to partition the input.
        sequence_parallel: If ``True`` the partial matmul outputs are combined with a
            sequence reduce-scatter (Megatron sequence parallelism) instead of an
            all-reduce: this sums the partial products across TP *and* shards the
            result along the sequence, leaving the residual stream sequence-sharded.
        axis_name: Mesh axis used for tensor parallelism.
    """

    def __init__(
        self,
        in_features: int,
        out_features: int,
        has_bias: bool = True,
        weight_init: Callable | None = None,
        bias_init: Callable | None = None,
        input_is_parallel: bool = False,
        sequence_parallel: bool = False,
        axis_name: str = "tp",
    ) -> None:
        super().__init__()

        if sequence_parallel and not input_is_parallel:
            # SP RowParallel always follows an SP ColumnParallel, whose output is
            # already feature-sharded. A non-parallel (replicated) SP input would
            # need a simultaneous sequence-gather + feature-scatter that Llama never
            # exercises; reject it rather than emit a subtly wrong graph.
            raise ValueError("RowParallelLinear: sequence_parallel requires input_is_parallel=True")

        self.in_features = in_features
        self.out_features = out_features
        # TODO: use ttnn.Tensor.tensor_topology() once CCL ops will properly update this property
        self.input_is_parallel = input_is_parallel
        self.sequence_parallel = sequence_parallel
        self.axis_name = axis_name
        self.cluster_axis = ttml.mesh().axis_index(axis_name)

        if weight_init is None:
            k = math.sqrt(1.0 / in_features)
            weight_init = ttml.init.uniform(-k, k)
        if bias_init is None:
            k = math.sqrt(1.0 / in_features)
            bias_init = ttml.init.uniform(-k, k)

        weight_shape = (1, 1, out_features, in_features)
        weight_mapper = ttml.mesh().axis_mapper(axis_name, tdim=3)
        self.weight = Parameter(weight_init(weight_shape, mapper=weight_mapper))

        if has_bias:
            bias_shape = (1, 1, 1, out_features)
            self.bias = Parameter(bias_init(bias_shape))
            if sequence_parallel:
                mark_sequence_parallel(self.bias)
        else:
            self.bias = None

    def forward(self, x):
        if not self.input_is_parallel:
            # Split the input along the feature dimension across TP devices.
            x = ttml.ops.distributed.scatter(x, 3, self.cluster_axis)
        x = ttml.ops.linear.linear(x, self.weight.tensor, None)
        x = row_parallel_output(x, self.cluster_axis, self.sequence_parallel, self.input_is_parallel)
        if self.bias is not None:
            x = ttml.ops.binary.add(x, self.bias.tensor)
        return x
